real/assembly.h

```diff
@@ -414,6 +414,49 @@ static __inline Word64 SAR64(Word64 x, int n)
 	return (x >> n);
 }
 
+#elif defined(__xtensa__)
+
+typedef long long Word64;
+
+static __inline Word64 MADD64(Word64 sum64, int x, int y)
+{
+    return (sum64 + ((long long)x * y));
+}
+
+static __inline int MULSHIFT32(int x, int y)
+{
+    /* important rules for smull RdLo, RdHi, Rm, Rs:
+     *     RdHi and Rm can't be the same register
+     *     RdLo and Rm can't be the same register
+     *     RdHi and RdLo can't be the same register
+     * Note: Rs determines early termination (leading sign bits) so if you want to specify
+     *   which operand is Rs, put it in the SECOND argument (y)
+     * For inline assembly, x and y are not assumed to be R0, R1 so it shouldn't matter
+     *   which one is returned. (If this were a function call, returning y (R1) would
+     *   require an extra "mov r0, r1")
+     */
+    int ret;
+    asm volatile ("mulsh %0, %1, %2" : "=r" (ret) : "r" (x), "r" (y));
```

Comment:
I think this implementation should be wrapped in `#if XCHAL_HAVE_MUL32_HIGH` .. `#endif`.
Not all Xtensa processor configurations have the `mulsh` instruction, and the above macro can be used to detect whether the instruction is supported. This macro will be defined to 1 or 0 in "xtensa/config/core-isa.h".

Reply:
@igrr should I also be including xtensa/config/core-isa.h or just the #if check?

and how did you spot this one? its a really good catch.

Reply:
@igrr I'm seeing a compile time error with these conditionals. Are these not always defined? I see them defined here:

```
components/xtensa/esp32/include/xtensa/config/core-isa.h:#define XCHAL_HAVE_MUL32_HIGH		1	/* MULUH/MULSH instructions */
components/xtensa/esp32s3/include/xtensa/config/core-isa.h:#define XCHAL_HAVE_MUL32_HIGH		1	/* MULUH/MULSH instructions */
components/xtensa/esp32s2/include/xtensa/config/core-isa.h:#define XCHAL_HAVE_MUL32_HIGH		1	/* MULUH/MULSH instructions */
```

but these don't look defined at build time:

```
/Users/cmorgan/projects/esp-box/examples/mp3_demo/managed_components/chmorgan__esp-libhelix-mp3/libhelix-mp3/real/assembly.h:447:2: error: #error Missing definition of MULSHIFT32
 #error Missing definition of MULSHIFT32
  ^~~~~
/Users/cmorgan/projects/esp-box/examples/mp3_demo/managed_components/chmorgan__esp-libhelix-mp3/libhelix-mp3/real/assembly.h:462:2: error: #error Missing definition of FASTABS
 #error Missing definition of FASTABS
  ^~~~~
/Users/cmorgan/projects/esp-box/examples/mp3_demo/managed_components/chmorgan__esp-libhelix-mp3/libhelix-mp3/real/dct32.c: In function 'xmp3_FDCT32':
/Users/cmorgan/projects/esp-box/examples/mp3_demo/managed_components/chmorgan__esp-libhelix-mp3/libhelix-mp3/real/dct32.c:109:26: error: implicit declaration of function 'MULSHIFT32' [-Werror=implicit-function-declaration]
     b0 = a0 + a3;   b3 = MULSHIFT32(*cptr++, a0 - a3) << (s0); \
                          ^~~~~~~~~~
/Users/cmorgan/projects/esp-box/examples/mp3_demo/managed_components/chmorgan__esp-libhelix-mp3/libhelix-mp3/real/dct32.c:164:2: note: in expansion of macro 'D32FP'
  D32FP(0, 1, 5, 1);
```

Reply:
@igrr? Please help!!! :-)

Reply:
> @igrr? Please help!!! :-)
HI，chmorgan, did you fixed it now? I find the dependencies esp-libhelix-mp3 updated to 1.0.2 now;

Reply:
I applied the changes @igrr suggested and the defines aren’t set anywhere. I’m hoping he replies and I can resolve. Otherwise I’ll have to revert, sorry for the breakage, I’ll fix tonight even if I don’t hear from him.

Reply:
ok, thanks a lot!

Reply:
@igrr thanks! That was my initial thought but wanted to confirm.

Reply:
@espressif2022 esp-libhelix-mp3 v1.0.3 is pushing up right now, should build correctly now with the instruction checks.